Fixed #24900 -- Allowed migrating backward to squashed migrations.

jacobtylerwalls opened the pull request:
ticket-24900

shaib reviewed (approved):
Looking good, I see one issue with this, but it's a little theoretical:

The loader is changed by the PR to be always in one of two modes; either it loads only the replacement migrations, or it loads only the replaced migrations.

This is acceptable for the commands, because they always specify at most on target migration.

But the executor API takes a list of targets. In principle, some of them could be squashed and some of them could be replacements. Before this PR, it can handle this case, after the PR it can't.

I don't see where this would be used -- I don't see where there would be multiple targets, except when the targets are all the last migrations, in which case this doesn't matter at all.

jacobtylerwalls commented:
> But the executor API takes a list of targets. In principle, some of them could be squashed and some of them could be replacements. Before this PR, it can handle this case, after the PR it can't.

I wondered about this as well; however, I believe my changes target only the cases that were failing with `KeyError` before. AFAICS, any case where we execute this new condition:

https://github.com/django/django/blob/d137340d86a6e84c447a836fec561d57b53b10c9/django/db/migrations/executor.py#L29

would have failed here with `KeyError` when reaching that target during `for target in targets`:

https://github.com/django/django/blob/d137340d86a6e84c447a836fec561d57b53b10c9/django/db/migrations/executor.py#L47-L54

Unless the squashed migration has not been applied (!), in which case we execute the `else` branch, but on main I see this failing anyway with `django.db.migrations.exceptions.NodeNotFoundError: Node ('migrations', '0001_initial') not a valid node`.

Do let me know if I'm wrong about this, and thanks very much for the review!

shaib commented:
I think your analysis is wrong, but, as I said above, the use-case is theoretical.

Before your changes, if the loader builds the graph with `replace_migrations` set to `False`, the graph will include all migrations -- both squashed and replacements. The executor (before your changes) never does that, so one would need to meddle with it -- change the flag in its internal loader object -- in order to get the behavior I'm talking about.

I think anyone who does that, probably deserves what they get.

But now, thinking again, there's an issue with this approach which does merit consideration.

Consider two apps: `app_a` and `app_b`, each with `0001_initial`, `0002_second`, `0003_third`, and `0001_squashed_0002_together` (with the obvious replacements); and where `app_b.0001_initial` depends on `app_a.0003_third`. Assume none of them are applied.

If I understand correctly, `manage.py migrate app_b 0001` will ignore `app_a.0001_squashed_0002_together` and execute app_a's 0001 and 0002 separately.

I think this can be considered acceptable, but needs to be documented because it is surprising behavior. At a glimpse, preventing it would probably require the change to be significantly more involved.

jacobtylerwalls commented:
> Before your changes, if the loader builds the graph with replace_migrations set to False, the graph will include all migrations -- both squashed and replacements. The executor (before your changes) never does that, so one would need to meddle with it -- change the flag in its internal loader object -- in order to get the behavior I'm talking about.

Ah, got it now, thanks.

Yes, seems a little funny to complain that replacement migrations won't be used when explicitly setting an undocumented `replace_migrations` attribute False. I agree with you this is theoretical and probably not worth much back and forth, but for anyone else following: I suspect the status quo in the scenario Shai describes--loading both squashed and replacements into the graph--was already broken: wouldn't the executor try to execute both 0001_squashed_0002_together and 0002_foobar and then crash with "not a valid node" (or "table already exists" if using the executor directly) or similar?

> If I understand correctly, manage.py migrate app_b 0001 will ignore app_a.0001_squashed_0002_together and execute app_a's 0001 and 0002 separately.

> I think this can be considered acceptable, but needs to be documented because it is surprising behavior. At a glimpse, preventing it would probably require the change to be significantly more involved.

Sure, let me think about the best way to put this. The hypothetical use case on the ticket was a production rollback, say, from `app_b.0001_squashed_0002_together` to `app_b.0001_initial`. For rollbacks, I think using more atomic migrations is probably the user-friendly way to go.

smithdc1 commented on docs/topics/migrations.txt:
```diff
@@ -715,6 +715,12 @@ You must then transition the squashed migration to a normal migration by:
+    Support for migrating to replaced migrations still present on disk was
+    added.
+    In this scenario, no squashed migration will be executed: only the
+    migrations they originally replaced, even for dependencies in other apps.
```
It's probably due to me not being that familiar with migrations, but I'm really struggling with what this means. I wonder if a small example could help?

jacobtylerwalls replied:
Yup, it would be better to avoid the term "squashed migration" -- which is the one that's squashed: the one that _got_ squashed or the one that _did_ the squashing? (The one that did the squashing gets labeled "...squashed..."!)

Open to the idea of a small example but don't want to go there just quite yet until I'm sure this change is acceptable. This doc has other problems, for instance: earlier it says it's okay to commingle replaced/replacing migrations, then below it says you are strongly encouraged to get rid of the old ones, and when doing so you "must" do the following to convert the "squashed" migration to "regular" etc. I'm just not sure this is the PR for a big rewrite. Speaking of which, we have #14408.

jacobtylerwalls replied:
Thankfully, this has been mooted by Mariusz's suggested smaller diff!

shaib reviewed (approved):
LGTM

felixxm commented on tests/migrations/test_executor.py:
```diff
@@ -104,6 +104,29 @@ def test_run_with_squashed(self):
+        self.assertTableExists('migrations_author')
+        self.assertTableNotExists('migrations_book')
+
+        # Migrate to zero.
+        executor.loader.build_graph()
+        executor.migrate([('migrations', None)])
+        self.assertTableNotExists('migrations_author')
+        self.assertTableNotExists('migrations_book')
```
This should be in a `finally` block.

felixxm reviewed (commented):
@jacobtylerwalls Thanks for this patch :+1: However, I'm not sure if rebuilding the migration graph multiple times is an acceptable solution to this issue :thinking:

I'm quite surprised that an extra rebuild, added to the `sqlmigrate`, is needed for `test_sqlmigrate_squashed_migration` which passes a squashed migration as a target, not a replaced migration :thinking: It's also worth noting that applying replaced migrations work without this patch, e.g.:
- create migration `0001_initial`
- create migration `0002_sth`
- squash migrations `0001` and `0002`
- migrate to `0001_initial`

so it's only an issue when reversing migrations.

Maybe it's enough to patch `migrate` when replaced (and already applied) node is used :thinking: e.g.
```diff
diff --git a/django/core/management/commands/migrate.py b/django/core/management/commands/migrate.py
index 34fcb9bc58..703383fc5e 100644
--- a/django/core/management/commands/migrate.py
+++ b/django/core/management/commands/migrate.py
@@ -143,12 +143,13 @@ class Command(BaseCommand):
                 target = (app_label, migration.name)
                 # Partially applied squashed migrations are not included in the
                 # graph, use the last replacement instead.
-                if (
-                    target not in executor.loader.graph.nodes and
-                    target in executor.loader.replacements
-                ):
-                    incomplete_migration = executor.loader.replacements[target]
-                    target = incomplete_migration.replaces[-1]
+                if target not in executor.loader.graph.nodes:
+                    if target in executor.loader.replacements:
+                        incomplete_migration = executor.loader.replacements[target]
+                        target = incomplete_migration.replaces[-1]
+                    else:
+                        executor.loader.replace_migrations = False
+                        executor.loader.build_graph()
                 targets = [target]
             target_app_labels_only = False
         elif options['app_label']:
```

jacobtylerwalls commented:
Thanks, I like that diff. Less invasive.

That's the branch we added in #14727  _after_ I opened this patch, so I'm glad we took things in this order! I've rebased. I also had to add another graph-reload branch in 800d0fc. (To compare complexity the original changes are [here](https://github.com/django/django/commit/d137340d86a6e84c447a836fec561d57b53b10c9).)

***

> I'm quite surprised that an extra rebuild, added to the sqlmigrate, is needed for test_sqlmigrate_squashed_migration which passes a squashed migration as a target, not a replaced migration 🤔

Right, that did seem noteworthy. The original solution did this, as Shai put it:

> The loader is changed by the PR to be always in one of two modes; either it loads only the replacement migrations, or it loads only the replaced migrations.

> Before your changes, if the loader builds the graph with replace_migrations set to False, the graph will include all migrations -- both squashed and replacements. 

And so we needed to patch `sqlmigrate`, which sets `replace_migrations` False purely to be permissive, not to limit the nodes of the graph -- which would be bad, since the test you mentioned passes a squashed migration:

https://github.com/django/django/blob/36ab69b867f36615b921d4686e92906de6be1feb/django/core/management/commands/sqlmigrate.py#L35-L37

But all to say, yes I like the less invasive diff better 👍🏻

felixxm commented on tests/migrations/test_executor.py:
```diff
@@ -104,6 +104,29 @@ def test_run_with_squashed(self):
+            executor = MigrationExecutor(connection)
+            executor.migrate([('migrations', None)])
+            self.assertTableNotExists('migrations_author')
+            self.assertTableNotExists('migrations_book')
```
I'd add a similar test to the `migrations.test_commands`:
```python
    @override_settings(MIGRATION_MODULES={'migrations': 'migrations.test_migrations_squashed'})
    def test_migrate_backward_to_replaced_migration(self):
        try:
            call_command('migrate', 'migrations', '0001_squashed_0002', verbosity=0)
            call_command('migrate', 'migrations', '0001_initial', verbosity=0)
            ...
        finally:
            call_command('migrate', 'migrations', 'zero', verbosity=0)
```

felixxm commented on django/core/management/commands/migrate.py:
```diff
@@ -141,14 +141,17 @@ def handle(self, *args, **options):
+                    # backward to a replaced migration.
+                    else:
+                        executor.loader.replace_migrations = False
+                        executor.loader.build_graph()
```
Changes in the `migrate` command are unnecessary when you added a similar fix to the executor.

felixxm commented on tests/migrations/test_loader.py:
```diff
@@ -492,6 +492,22 @@ def test_loading_squashed_ref_squashed(self):
         with self.assertRaisesMessage(NodeNotFoundError, msg):
             loader.graph.forwards_plan(('app1', '3_auto'))
 
+        # Possible if loader.replace_migrations is False.
+        # MigrationExecutor.migration_plan() is responsible for setting it False.
+        loader.replace_migrations = False
+        loader.build_graph()
+        plan = set(loader.graph.forwards_plan(('app1', '3_auto')))
+        plan = plan - loader.applied_migrations.keys()
+        expected_plan = {
+            ('app1', '1_auto'),
+            ('app2', '1_auto'),
+            ('app2', '2_auto'),
+            ('app1', '2_auto'),
+            ('app1', '3_auto'),
+        }
+        self.assertEqual(plan, expected_plan)
+        loader.replace_migrations = True
+
```
We didn't make any changes in the loader and this assertions work without a patch. Please move them to the first commit.

felixxm commented on tests/migrations/test_executor.py:
```diff
@@ -104,6 +104,29 @@ def test_run_with_squashed(self):
+    @override_settings(MIGRATION_MODULES={'migrations': 'migrations.test_migrations_squashed'})
+    def test_migrate_to_replaced_migration(self):
+        executor = MigrationExecutor(connection)
+
```
Chop blank line.

felixxm commented on django/db/migrations/executor.py:
```diff
@@ -44,6 +44,12 @@ def migration_plan(self, targets, clean_start=False):
                 # may roll back dependencies in other apps that don't need to
                 # be rolled back); instead roll back through target's immediate
                 # child(ren) in the same app, and no further.
+                if target not in self.loader.graph.node_map and self.loader.replace_migrations:
```
The above comment should be moved below this block.

felixxm reviewed (commented):
@jacobtylerwalls Thanks for updates :+1: 

@shaib What do you think about the new version?

felixxm commented:
@jacobtylerwalls Thanks for updates :+1:

vanschelven commented:
[https://code.djangoproject.com/ticket/36168](I opened a new issue on the Django issue tracker that's likely related to this one, would love to hear your thoughts)
